### PR TITLE
[WIP] [3.3.x] Implement logged in users limit

### PR DESCRIPTION
**Pull request by Swoq**

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/security/auth/rest/RestAwareAuthenticationSuccessHandler.java [hunk 1]
```diff
@@ -40,22 +42,26 @@
...
-        this.refreshTokenRepository = refreshTokenRepository;
+        this.loggedInUsersLimitService = loggedInUsersLimitService;
     }
 
     @Override
     public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response,
                                         Authentication authentication) throws IOException, ServletException {
         SecurityUser securityUser = (SecurityUser) authentication.getPrincipal();
+        if (loggedInUsersLimitService.isOverLimit(securityUser.getId()))
+            throw new AuthenticationServiceException("Violation of logged in users amount limit");
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/security/auth/rest/RestAwareAuthenticationSuccessHandler.java:59]: Please check if this exception is expected in this handler method

File: application/src/main/java/org/thingsboard/server/service/security/auth/oauth2/Oauth2AuthenticationSuccessHandler.java [hunk 2]
```diff
@@ -95,8 +93,9 @@ public void onAuthenticationSuccess(HttpServletRequest request,
             SecurityUser securityUser = mapper.getOrCreateUserByClientPrincipal(request, token, oAuth2AuthorizedClient.getAccessToken().getTokenValue(),
                     registration);
 
-            JwtToken accessToken = tokenFactory.createAccessJwtToken(securityUser);
-            JwtToken refreshToken = refreshTokenRepository.requestRefreshToken(securityUser);
+            Pair<JwtToken, JwtToken> tokensPair = tokenFactory.getAccessAndRefreshTokens(securityUser);
+            JwtToken accessToken = tokensPair.getFirst();
+            JwtToken refreshToken = tokensPair.getSecond();
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/security/auth/oauth2/Oauth2AuthenticationSuccessHandler.java:99]: Why don't you check the limits?

File: application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java [hunk 3]
```diff
@@ -49,34 +51,51 @@ protected void initCache() {
...
+        return isTokenExpired || isUserExpired;
+    }
+
+    public void outdateTokens(String id) {
+        tokenOutdatageTimeCache.put(id, System.currentTimeMillis());
+    }
+
+    public void outdateTokens(RawAccessJwtToken token) {
+        Jws<Claims> claimsJws = tokenFactory.parseTokenClaims(token);
+        Optional.of(claimsJws.getBody().get("tokenId").toString()).ifPresent(this::outdateTokens);
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java:82]: You have TOKEN_ID constant in JwtTokenFactory, please use it

Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java:82]: tokenId claim might not be present in the token

File: application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java [hunk 4]
```diff
@@ -49,34 +51,51 @@ protected void initCache() {
...
-                         * as all the tokens issued before the outdatage time
-                         * are now expired by themselves
-                         * */
-                        tokenOutdatageTimeCache.evict(toKey(userId));
-                        return false;
-                    }
-                })
-                .orElse(false);
+        Jws<Claims> claimsJws = tokenFactory.parseTokenClaims(token);
+        Optional<String> tokenId = Optional.ofNullable(claimsJws.getBody().get("tokenId").toString());
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java:62]: `claimsJws.getBody().get("tokenId")` might be null, not `claimsJws.getBody().get("tokenId").toString()`

File: application/src/main/java/org/thingsboard/server/service/security/auth/LoggedInUsersLimitService.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        String key = toKey(userId);
+        if (currentAmount.isPresent())
+            loginLimitCache.put(key, currentAmount.get() + 1);
+        else
+            loginLimitCache.put(key, 1L);
+    }
+
+    public Optional<Long> getCurrentAmount(UserId userId) {
+        return Optional.ofNullable(loginLimitCache.get(toKey(userId), Long.class));
+    }
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/security/auth/LoggedInUsersLimitService.java:72]: "Logged in users" does not reflect what this feature actually does. It is rather an amount of active sessions per one user. Please refactor wherever possible

File: application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java (hunk 4 above)
Comment (Swoq) [reply application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java:62]: resolved

File: application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java (hunk 3 above)
Comment (Swoq) [reply application/src/main/java/org/thingsboard/server/service/security/auth/TokenOutdatingService.java:82]: that's a good point. resolved

**Pull request closed without merging**